sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -274,38 +274,40 @@ object JavaTypeInference {
 
       case c if listType.isAssignableFrom(typeToken) =>
         val et = elementType(typeToken)
-        val newTypePath = s"""- array element class: "${et.getType.getTypeName}"""" +:
-          walkedTypePath
+        walkedTypePath.recordArray(et.getType.getTypeName)
         val (dataType, elementNullable) = inferDataType(et)
         val mapFunction: Expression => Expression = element => {
           // upcast the array element to the data type the encoder expected.
           deserializerForWithNullSafetyAndUpcast(
             element,
             dataType,
             nullable = elementNullable,
-            newTypePath,
+            walkedTypePath,
             (casted, typePath) => deserializerFor(et, casted, typePath))
         }
 
         UnresolvedMapObjects(mapFunction, path, customCollectionCls = Some(c))
 
       case _ if mapType.isAssignableFrom(typeToken) =>
         val (keyType, valueType) = mapKeyValueType(typeToken)
-        val newTypePath = (s"""- map key class: "${keyType.getType.getTypeName}"""" +
-          s""", value class: "${valueType.getType.getTypeName}"""") +: walkedTypePath
+        walkedTypePath.recordMap(keyType.getType.getTypeName,
+          valueType.getType.getTypeName)
+
+        val newTypePathForKey = walkedTypePath.copy()
+        val newTypePathForValue = walkedTypePath.copy()
```

Comment:
Sorry for the back and forth. But seems it's better to make `WalkedTypePath` immutable as there are branches. It's hard to maintain and we can easily mess it up if we forget the call `copy` somewhere.

Reply:
Yeah, same understanding. I'll revert back to let WalkedTypePath be immutable one.